### PR TITLE
Fuzzers: 007-timing: add routing BELs fuzzer

### LINKED ISSUE #890: SDF output is missing mux timing
The current BEL timing fuzzer is not documenting BEL routing muxes, e.g. [ABCD]OUTMUX.  For example F7A -> AOUT is at ``bel_d_sm_f7amux_amux``.

### PR DESCRIPTION
**Pull request by kgugala**
fixes #890

### CODE REVIEW

**Comment by kgugala** [comment]
We need https://github.com/SymbiFlow/python-sdf-timing/issues/32 before we can proceed with this one

File: fuzzers/007-timing/bel/Makefile [hunk 1]
```diff
@@ -19,15 +19,5 @@ build/bel_timings.json: build/fixup_timings
 
 build/sdf: build/bel_timings.json
```
Comment (litghost) [inline fuzzers/007-timing/bel/Makefile]: .ok?

File: fuzzers/007-timing/bel/Makefile [hunk 2]
```diff
@@ -1,4 +1,4 @@
-all: pushdb
+all: build/sdf
```
Comment (litghost) [inline fuzzers/007-timing/bel/Makefile]: .ok?

**Comment by litghost** [comment]
Also should these annotations be "DEVICE", or maybe "COND" or "INTERCONNECT"?

**Comment by litghost** [comment]
From looking through the SDF spec, it seems like "INTERCONNECT" is the right description, and will allow for a good structure as the "INTERCONNECT" item also describing both an out port (e.g. B6LUT) and an in port (e.g. BMUX, etc).

**Comment by kgugala** [comment]
> Also should these annotations be "DEVICE", or maybe "COND" or "INTERCONNECT"?

DEVICE is suppose do be used for describing delay of a black box. From the spec:

`The DEVICE keyword represents the intrinsic delay of a module or gate. Intrinsic delay is specific to the type of the object and so has the same value for every instance of that module or gate. Conceptually, this represents all path delays through the object, independent of loading or input slope`

COND us used for conditional timings. INTERCONNECT is something that may fit here, but it requires providing input and output. That would be hard to extract from timing models Vivado reports for the routing bels

**Comment by litghost** [comment]
> > Also should these annotations be "DEVICE", or maybe "COND" or "INTERCONNECT"?
> INTERCONNECT is something that may fit here, but it requires providing input and output. That would be hard to extract from timing models Vivado reports for the routing bels

But it is present.  ``sl_b6lut_bmux`` is ``b6lut`` to ``bmux`` routing delay.  There also turns out to be a ``bmuxl`` delay, but it appears to be equal to the ``bmux`` delay?

**Comment by kgugala** [comment]
> > > Also should these annotations be "DEVICE", or maybe "COND" or "INTERCONNECT"?
> > > INTERCONNECT is something that may fit here, but it requires providing input and output. That would be hard to extract from timing models Vivado reports for the routing bels
> 
> But it is present. `sl_b6lut_bmux` is `b6lut` to `bmux` routing delay. There also turns out to be a `bmuxl` delay, but it appears to be equal to the `bmux` delay?

sure I can change it to INTERCONNECT, what about e.g. `sl_d6lut_d5lut_diff`? what do you think is the input and output?

yep, `Xmuxl` delays seem to be equal to `Xmux` ones

**Comment by litghost** [comment]
``sl_d6lut_d5lut_diff`` does seem different.  I'm not even sure where it would go?

**Comment by kgugala** [comment]
those are the timings I'm not sure how to deal with:
```
                (DEVICE sl_a6lut_a5lut_diff (0.003::0.004)(0.022::0.027))
                (DEVICE sl_a6lut_int_diff (0.010::0.012)(0.019::0.024))
                (DEVICE sl_b6lut_b5lut_diff (0.004::0.005)(0.022::0.027))
                (DEVICE sl_b6lut_int_diff (0.009::0.011)(0.018::0.022))
                (DEVICE sl_c6lut_c5lut_diff (0.006::0.007)(0.024::0.029))
                (DEVICE sl_c6lut_int_diff (0.009::0.011)(0.018::0.022))
                (DEVICE sl_d6lut_d5lut_diff (0.004::0.005)(0.021::0.026))
                (DEVICE sl_d6lut_int_diff (0.009::0.011)(0.017::0.021))
```

The rest is pretty straight forward

**Comment by litghost** [comment]
> those are the timings I'm not sure how to deal with:

Given that we don't even know where they would end up in the pb_type, let's just output them as "DEVICE" and the others as "INTERCONNECT".  Will that work?

**Comment by kgugala** [comment]
Sure, let's do it like that for now

File: fuzzers/007-timing/bel/Makefile (hunk 1 above)
Comment (kgugala) [reply fuzzers/007-timing/bel/Makefile]: fixed

File: fuzzers/007-timing/bel/Makefile (hunk 2 above)
Comment (kgugala) [reply fuzzers/007-timing/bel/Makefile]: fixed

**Comment by kgugala** [comment]
The timings are now emitted as INTERCONN (except the diffs)

I have grouped all the routing bels timings in a single cell_type named "ROUTING_BEL"

**Comment by kgugala** [comment]
why did kokoro fail? I cannot access the log. @litghost can you check it?

**Comment by kgugala** [comment]
kokoro force-run helped

**Comment by kgugala** [comment]
hmm, It looks like CI failed because of Vivado bug:

```
007-timing               - 9m29s: TclStackFree: incorrect freePtr. Call out of sequence?^M
007-timing               - 9m29s: make[2]: *** [build/slicem.txt] Error 134^M
```
I'll rerun the tests

**Comment by kgugala** [comment]
@litghost CI is green, can you look at the generated SDFs?

**Comment by mithro** [comment]
```sdf
(CELL
    (CELLTYPE "RAMBFIFO36E1")
    (INSTANCE RAMBFIFO36E1)
    (TIMINGCHECK
        (HOLD ADDRA15L (posedge CLKARDCLKU) (-0.515::0.357))
        (SETUP ADDRA15L (posedge CLKARDCLKU) (-0.357::0.515))
        (HOLD ADDRAU (posedge CLKARDCLKU) (-0.566::0.360))
        (SETUP ADDRAU (posedge CLKARDCLKU) (-0.360::0.566))
        (HOLD ADDRB15L (posedge CLKBWRCLKU) (-0.515::0.357))
        (SETUP ADDRB15L (posedge CLKBWRCLKU) (-0.357::0.515))
        (HOLD ADDRBU (posedge CLKBWRCLKU) (-0.566::0.360))
        (SETUP ADDRBU (posedge CLKBWRCLKU) (-0.360::0.566))
        (HOLD WEAU (posedge CLKARDCLKU) (-0.532::0.197))
        (SETUP WEAU (posedge CLKARDCLKU) (-0.197::0.532))
        (HOLD WEBU (posedge CLKBWRCLKU) (-0.532::0.197))
        (SETUP WEBU (posedge CLKBWRCLKU) (-0.197::0.532))
    )
)
(CELL
    (CELLTYPE "RAMBFIFO36E1_DOB_REG_L_1_DOB_REG_U_1")
    (INSTANCE RAMBFIFO36E1)
    (TIMINGCHECK
        (HOLD REGCEBU (posedge REGCLKBU) (-0.360::0.155))
        (SETUP REGCEBU (posedge REGCLKBU) (-0.155::0.360))
    )
)
```

**Comment by mithro** [comment]
```sdf
(CELL
    (CELLTYPE "REG_INIT_LAT_QH")
    (INSTANCE SLICEL)
    (DELAY
        (ABSOLUTE
            (IOPATH SR Q (0.164::0.204)(0.487::0.604))
        )
    )
)
(CELL
    (CELLTYPE "REG_INIT_LAT_QL")
    (INSTANCE SLICEL)
    (DELAY
        (ABSOLUTE
            (IOPATH SR Q (0.198::0.246)(0.638::0.791))
        )
    )
)
```

**Comment by mithro** [comment]
LGTM.

**Comment by kgugala** [comment]
@litghost is there anything else I should fix here?

**Comment by litghost** [comment]
Why are we writing out the interconnect's without grouping?

```
+        (DELAY
+            (ABSOLUTE
+                (INTERCONNECT o6 amux (0.061::0.076)(0.147::0.182))
+            )
+        )
```

and

```
+        (DELAY
+            (ABSOLUTE
+                (INTERCONNECT o5 amux (0.063::0.078)(0.143::0.178))
+            )
+        )
```

Should just be:
```
+        (DELAY
+            (ABSOLUTE
+                (INTERCONNECT o5 amux (0.063::0.078)(0.143::0.178))
+                (INTERCONNECT o6 amux (0.061::0.076)(0.147::0.182))
+            )
+        )
```

**Comment by kgugala** [comment]
> Better grouping on INTERCONNECT output?

This requires change in the python-sdf-parser. It makes sense to add the grouping option. I think we should group the timings by CELLTYPE and INSTANCE (as it is for regular BELs). Timings sorting should handle the rest.

**Comment by litghost** [comment]
> > Better grouping on INTERCONNECT output?
> 
> This requires change in the python-sdf-parser. It makes sense to add the grouping option. I think we should group the timings by CELLTYPE and INSTANCE (as it is for regular BELs). Timings sorting should handle the rest.

I believe it is important to get this group so it is easier to spot check the import.  Let's update python-sdf-parser?

**Comment by kgugala** [comment]
> > > Better grouping on INTERCONNECT output?
> > 
> > 
> > This requires change in the python-sdf-parser. It makes sense to add the grouping option. I think we should group the timings by CELLTYPE and INSTANCE (as it is for regular BELs). Timings sorting should handle the rest.
> 
> I believe it is important to get this group so it is easier to spot check the import. Let's update python-sdf-parser?

I have fixed that in `python-sdf-parser` (https://github.com/SymbiFlow/python-sdf-timing/pull/35).
When the PR is merged there I'll bump it here

**Comment by kgugala** [comment]
@litghost I have bumped the `python-sdf-parser` submodule. The timings should now be grouped

**Comment by litghost** [comment]
> @litghost I have bumped the `python-sdf-parser` submodule. The timings should now be grouped

I'll take a look when CI finished, thanks!

**Comment by kgugala** [comment]
@litghost i have rebased the PR. CI is green. Please take a look on the sdfs

**Comment by litghost** [comment]
> @litghost i have rebased the PR. CI is green. Please take a look on the sdfs

I think this is better, but not ideal.

I think the INTERCONNECT group should be by output.  I can see this being done a in a couple different ways.  Two seems good:
1. Make INTERCONNECT groups by output, so a group of ``a5ff`` and a group of ``b5ff``, etc.
2. Sort the INTERCONNECT groups by output, follow by input signal

Thoughts?

**Comment by litghost** [comment]
007 failed on CI?

**Comment by kgugala** [comment]
> 007 failed on CI?

yep, It seems to be a bug in Makefile running multiple Vivado instances in the same directory. Should be fixed now. Let's wait for the CI

**Comment by kgugala** [comment]
@litghost the CI is green now. Please take a look

**Pull request merged**